### ISSUE TITLE
Warning on heading element

### ISSUE THREAD

**Issue by shimaokk** [validator/validator#1572]
I get an “Empty heading” warning even though the `<h1>-<h6>` have a child elements have an aria-label or aria-labelledby attribute.

**Example 1**
```html
<h1>
  <svg role="img" aria-label="Example">
    ...
  </svg>
</h1>
```

**Example 2**
```html
<h1>
  <button type="button" aria-labelledby="example"></button>
</h1>
<div id="example">Example</div>
```

**Comment by yoruaki** [comment]
Hello everyone. I'm the one who originally submitted this issue.

After much discussion and learning, I now understand that this warning is due to a difference in how HTML semantics and ARIA roles are interpreted.

This distinction is clearly seen when comparing it to the `alt` attribute on an `<img>` element. The `alt` attribute is part of the **HTML specification itself** and is recognized by validators as a valid alternative for the image content. That's why a warning isn't issued for `<img src="..." alt="text">`.

However, the `<h1>` tag is based on a stricter semantic rule that it must contain **human-readable text** inside. `aria-label` is intended for assistive technologies and isn't considered by the validator as content that "fills" the `<h1>` tag.

I've come to understand that this warning isn't a bug, but rather an intentional behavior designed to encourage correct HTML structuring.

Moving forward, I'll be including invisible text inside the `<h1>` tag to resolve this. This approach satisfies both the validator's requirements and accessibility standards.

```html
<h1>
    <svg role="img" aria-label="Logo alternative text">...</svg>
    <span class="sr-only">Logo alternative text</span>
</h1>
```

Thank you to everyone who took an interest in this issue. It has been a valuable learning experience for me.

**Comment by scottaohara** [comment]
your code example would create a heading with the text:
"logo alternative text logo alternative text"

also, this warning isn't due to the use of ARIA - you'd get the same error if there were no ARIA used on the svg as well.

if you, however, did something like:

```
<h1>
  <svg role=img aria-labelledby=f>
    <title id=f hidden>alternative text</title>  
    <!-- or use a <text> element instead if you don't want the title's tooltip to appear on hover -->
  </svg>
</h1>
```

then you'd get the correct heading text (only a single instance of the alternative text) and pass the validator.

**Comment by sideshowbarker** [comment]
Closed per https://github.com/validator/validator/issues/1572#issuecomment-3401722983
